src/hdmf/common/resources.py

```diff
@@ -253,15 +292,21 @@ def add_ref(self, **kwargs):
         container = kwargs['container']
         field = kwargs['field']
         key = kwargs['key']
-        resource_name = kwargs['resource_name']
-        resource_id = kwargs['entity_id']
-        resource_uri = kwargs['entity_uri']
+        entity_id = kwargs['entity_id']
+        entity_uri = kwargs['entity_uri']
         add_rsc = False
+        resource_table_idx = kwargs['resource_table_idx']
+        if resource_table_idx is not None:
+            resource_table_idx = kwargs['resource_table_idx']
+        else:
+            resource_name = kwargs['resource_name']
+            resource_uri = kwargs['resource_uri']
+            resource_table_idx = self.add_resource(resource_name, resource_uri)
 
-        if resource_name is not None and resource_id is not None and resource_uri is not None:
+        if ((resource_table_idx or resource_name) is not None and entity_id is not None and entity_uri is not None):
             add_rsc = True
-        elif not (resource_name is None and resource_id is None and resource_uri is None):
-            msg = ("Specify all or none of resource_name, entity_id, and entity_uri arguments. "
+        elif not ((resource_table_idx and resource_name is None) and entity_id is None and resource_uri is None):
```

Comment:
I believe based on the new logic above, `resource_table_idx` will never be None at this point. If `resource_table_idx` is None and `resource_name` is None, then the call to `self.add_resource` on line 304 will raise a error from the docval check saying that the name argument is missing.

I believe these checks should be split up in the following way: Above line 299, there should be a check that says if `resource_table_idx` is not None, then if `resource_name` is not None or `resource_table_idx` is not None, then raise an error. And if `resource_table_idx` is None, then either both `resource_name` and `resource_table_idx` are provided (if only one is provided, raise an error) or neither is provided. 

Then after resource_table_idx is possibly set, you can change line 306 to be:
`if (resource_table_idx is not None and entity_id is not None and entity_uri is not None):`
and line 308 to be:
`elif not (resource_table_idx is None and entity_id is None and resource_uri is None):`

The `add_rsc` flag variable should be renamed `add_entity` to reflect its new meaning (see line 339)

Please also add a test to ensure it is possible to run `add_ref` without any of the last five arguments.